File: Sources/Domain/Models/AIProvider.swift
import Foundation

/// Represents an AI coding assistant provider that can be monitored.
/// This is a rich domain model that encapsulates provider-specific behavior.
public enum AIProvider: String, CaseIterable, Sendable, Equatable, Hashable {
    case claude
    case codex
    case gemini

    /// Human-readable display name
    public var name: String {
        switch self {
        case .claude: "Claude"
        case .codex: "Codex"
        case .gemini: "Gemini"
        }
    }

    /// CLI command used to invoke this provider
    public var cliCommand: String {
        rawValue
    }

    /// Whether this provider is enabled by default
    public var isEnabled: Bool {
        true
    }

    /// URL to the provider's usage dashboard
    public var dashboardURL: URL? {
        switch self {
        case .claude:
            URL(string: "https://console.anthropic.com/settings/billing")
        case .codex:
            URL(string: "https://chatgpt.com/codex/settings/usage")
        case .gemini:
            URL(string: "https://gemini.google.com")
        }
    }

    /// URL to check the provider's service status
    public var statusPageURL: URL? {
        switch self {
        case .claude:
            URL(string: "https://status.claude.com/")
        case .codex:
            URL(string: "https://status.openai.com/")
        case .gemini:
            URL(string: "https://aistudio.google.com/status")
        }
    }

    /// Label for the session-based quota
    public var sessionQuotaLabel: String {
        switch self {
        case .claude, .codex: "Session"
        case .gemini: "Pro"
        }
    }

    /// Label for the weekly quota
    public var weeklyQuotaLabel: String {
        switch self {
        case .claude, .codex: "Weekly"
        case .gemini: "Flash"
        }
    }

    /// Whether this provider supports model-specific limits (e.g., Opus for Claude)
    public var supportsModelSpecificLimits: Bool {
        switch self {
        case .claude: true
        case .codex, .gemini: false
        }
    }
}

File: shared/domain/src/commonMain/kotlin/com/tddworks/claudebar/domain/provider/AIProvider.kt
package com.tddworks.claudebar.domain.provider

import com.tddworks.claudebar.domain.model.UsageSnapshot
import kotlinx.coroutines.flow.StateFlow

/**
 * Interface defining what an AI provider is.
 * Each provider (Claude, Codex, Gemini) is a rich domain model implementing this interface.
 * Providers expose their state via StateFlows for reactive UI consumption.
 */
interface AIProvider {
    // MARK: - Identity

    /** Unique identifier for the provider (e.g., "claude", "codex", "gemini") */
    val id: String

    /** Display name for the provider (e.g., "Claude", "Codex", "Gemini") */
    val name: String

    /** CLI command used to invoke the provider */
    val cliCommand: String

    /** URL to the provider's usage/billing dashboard */
    val dashboardURL: String?

    /** URL to the provider's status page */
    val statusPageURL: String?
        get() = null

    /** Whether the provider is enabled (user can toggle this) */
    var isEnabled: Boolean

    // MARK: - State (Observable via StateFlow)

    /** Whether the provider is currently syncing data */
    val isSyncing: StateFlow<Boolean>

    /** The current usage snapshot (null if never refreshed or unavailable) */
    val snapshot: StateFlow<UsageSnapshot?>

    /** The last error that occurred during refresh */
    val lastError: StateFlow<Throwable?>

    // MARK: - Operations

    /**
     * Checks if the provider is available (CLI installed, credentials present, etc.)
     */
    suspend fun isAvailable(): Boolean

    /**
     * Refreshes the usage data and updates the snapshot.
     */
    suspend fun refresh(): UsageSnapshot
}
